src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp

```diff
@@ -531,53 +519,38 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
-            _backgroundBrush = wil::unique_hbrush(CreateSolidBrush(color));
-
-            RECT windowRect = {};
-            ::GetWindowRect(_window.get(), &windowRect);
-            const auto cx = windowRect.right - windowRect.left;
-            const auto cy = windowRect.bottom - windowRect.top;
-
-            // Fill in ONLY the titlebar area. If we paint the _entirety_ of the
-            // window rect here, the single pixel of the bottom border (set in
-            // _UpdateFrameMargins) will be drawn, and blend with whatever the
-            // border color is.
-            RECT dragBarRect = GetDragAreaRect();
-            const auto dragHeight = RECT_HEIGHT(&dragBarRect);
-            dragBarRect.left = 0;
-            dragBarRect.right = cx;
-            dragBarRect.top = 0;
-            dragBarRect.bottom = dragHeight + yPos;
-            ::FillRect(hdc.get(), &dragBarRect, _backgroundBrush.get());
-
-            // Draw the top window border
-            RECT clientRect = { 0, 0, cx, yPos };
-            ::FillRect(hdc.get(), &clientRect, _backgroundBrush.get());
-
-            // Draw the left window border
-            clientRect = { 0, 0, xPos, cy };
-            ::FillRect(hdc.get(), &clientRect, _backgroundBrush.get());
-
-            // Draw the bottom window border
-            clientRect = { 0, cy - yPos, cx, cy };
-            ::FillRect(hdc.get(), &clientRect, _backgroundBrush.get());
-
-            // Draw the right window border
-            clientRect = { cx - xPos, 0, cx, cy };
-            ::FillRect(hdc.get(), &clientRect, _backgroundBrush.get());
+            const auto topBorderHeight = GetTopBorderHeight();
+
+            RECT rcTopBorder = ps.rcPaint;
+            rcTopBorder.top = 0;
+            rcTopBorder.bottom = topBorderHeight;
+            ::FillRect(hdc.get(), &rcTopBorder, GetStockBrush(BLACK_BRUSH));
+
+            if (ps.rcPaint.bottom > topBorderHeight)
+            {
+                // Create brush for borders, titlebar color.
+                const auto backgroundBrush = _titlebar.Background();
+                const auto backgroundSolidBrush = backgroundBrush.as<Media::SolidColorBrush>();
+                const auto backgroundColor = backgroundSolidBrush.Color();
+                const auto color = RGB(backgroundColor.R, backgroundColor.G, backgroundColor.B);
+                _backgroundBrush = wil::unique_hbrush(CreateSolidBrush(color));
+
+                RECT rcRest = ps.rcPaint;
+                rcRest.top = topBorderHeight;
+
+                // To paint on top of the original title bar, we have to set
+                // alpha channel to 255. This is a hack to do it with GDI.
+                HDC opaqueDc;
+                BP_PAINTPARAMS params = { sizeof(params), BPPF_NOCLIP | BPPF_ERASE };
+                HPAINTBUFFER buf = BeginBufferedPaint(hdc.get(), &rcRest, BPBF_TOPDOWNDIB, &params, &opaqueDc);

```

Comment:
I don't really know how GDI works so I found the first solution on Google to do this but maybe there is a better solution.